store/src/java/com/zimbra/cs/event/logger/EventLogger.java

```diff
@@ -4,49 +4,71 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
 
 public class EventLogger {
-    private static EventLogger eventLogger = null;
-    private final List<EventLogHandler> eventLogHandlers = new ArrayList<>();
+    private static final EventLogger eventLogger = new EventLogger();
+    private static final CopyOnWriteArrayList<EventLogHandler> eventLogHandlers = new CopyOnWriteArrayList<>();
+    private static final LinkedBlockingQueue<Event> eventQueue = new LinkedBlockingQueue<>();
+    private ExecutorService executorService = Executors.newFixedThreadPool(2);
 
-    private EventLogger() {}
+    private EventLogger() {
+        executorService.submit(new EventNotifier());
+    }
 
     public static EventLogger getEventLogger() {
-        if(eventLogger == null) {
-            eventLogger = new EventLogger();
-        }
         return eventLogger;
     }
 
     public void registerEventLogHandler(EventLogHandler logHandler) {
-        if(eventLogHandlers.contains(logHandler)) {
+        boolean present = eventLogHandlers.addIfAbsent(logHandler);
+        if(!present) {
             ZimbraLog.event.warn("Event Log Handler already registered %s", logHandler);
         }
-        else {
-            eventLogHandlers.add(logHandler);
-        }
     }
 
     public boolean unregisterEventLogHandler(EventLogHandler logHandler) {
-        if(eventLogHandlers.contains(logHandler)) {
-            return eventLogHandlers.remove(logHandler);
-        }
-        else {
+        boolean removed = eventLogHandlers.remove(logHandler);
+        if(!removed) {
             ZimbraLog.event.warn("Event Log Handler is not registered %s", logHandler);
-            return false;
         }
+        return removed;
     }
 
     @VisibleForTesting
     public void unregisterAllEventLogHandlers() {
         eventLogHandlers.clear();
     }
 
-    public void log(Event event) {
-        for (EventLogHandler eventLogHandler : eventLogHandlers) {
-            eventLogHandler.log(event);
+    public boolean log(Event event) {
+        try {
+            return eventQueue.add(event);
+        } catch (IllegalStateException e) {
+            ZimbraLog.event.debug("unable to add item for account %s to indexing queue", event.toString());
+            return false;
+        }
+    }
+
+    private static class EventNotifier implements Runnable {
+
+        @Override
+        public void run() {
+            while (true) {
+                try {
+                    Event event = eventQueue.take();
+                    Iterator<EventLogHandler> it = eventLogHandlers.iterator();
+                    while (it.hasNext()) {
+                        EventLogHandler logHandler = it.next();
+                        logHandler.log(event);
+                    }
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
```

Comment:
Maybe we should have two log categories - "events" for regular event-related messages and errors, and "eventlog" for the actual separate log of successful events?